README.md
# MY9291 LED driver library for Arduino ESP8266

Arduino library for the Taiwan Mingyang (MY-Semi) MY9291 LED driver. Based in the C driver by Maike Labs, included in the [Noduino SDK](https://github.com/icamgo/noduino-sdk).

Tested with Arduino Core for ESP8266.

## Example

A simple example, turning the red LEDs on.

```
#include <my9291.h>

#define MY9291_DI_PIN   13
#define MY9291_DCKI_PIN 15
#define CHANNELS        4

my9291 _my9291 = my9291(MY9291_DI_PIN, MY9291_DCKI_PIN, MY9291_COMMAND_DEFAULT, CHANNELS);

void setup() {
    _my9291.setColor((my9291_color_t) { 255, 0, 0, 0, 0 }); // RED 100% duty cycle (even if only 4 channels color has 5 components)
    _my9291.setState(true);
}

void loop() {}

```

For more in-depth examples please check the examples folder.


CHANGELOG.md
# DebounceEvent change log

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [2.0.0] 2017-08-26
### Added
- Support for 5 channels bulbs

### Changed
- Color has now 5 components, check examples

## [1.0.1] 2017-02-26
### Added
- Support for AVR architecture (untested)

## [1.0.0] 2017-02-25
- Initial version of the library


src/my9291.h
/*

MY9291 LED Driver Arduino library 2.0.0

Copyright (c) 2016 - 2026 MaiKe Labs
Copyright (C) 2017 - Xose Pérez

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#ifndef _my9291_h
#define _my9291_h

#include <Arduino.h>

#ifdef DEBUG_MY9291
    #if ARDUINO_ARCH_ESP8266
        #define DEBUG_MSG_MY9291(...) DEBUG_MY9291.printf( __VA_ARGS__ )
    #elif ARDUINO_ARCH_AVR
        #define DEBUG_MSG_MY9291(...) { char buffer[80]; snprintf(buffer, sizeof(buffer),  __VA_ARGS__ ); DEBUG_MY9291.print(buffer); }
    #endif
#else
    #define DEBUG_MSG_MY9291(...)
#endif

typedef enum my9291_cmd_one_shot_t {
        MY9291_CMD_ONE_SHOT_DISABLE = 0X00,
        MY9291_CMD_ONE_SHOT_ENFORCE = 0X01,
} my9291_cmd_one_shot_t;

typedef enum my9291_cmd_reaction_t {
        MY9291_CMD_REACTION_FAST = 0X00,
        MY9291_CMD_REACTION_SLOW = 0X01,
} my9291_cmd_reaction_t;

typedef enum my9291_cmd_bit_width_t {
        MY9291_CMD_BIT_WIDTH_16 = 0X00,
        MY9291_CMD_BIT_WIDTH_14 = 0X01,
        MY9291_CMD_BIT_WIDTH_12 = 0X02,
        MY9291_CMD_BIT_WIDTH_8 = 0X03,
} my9291_cmd_bit_width_t;

typedef enum my9291_cmd_frequency_t {
        MY9291_CMD_FREQUENCY_DIVIDE_1 = 0X00,
        MY9291_CMD_FREQUENCY_DIVIDE_4 = 0X01,
        MY9291_CMD_FREQUENCY_DIVIDE_16 = 0X02,
        MY9291_CMD_FREQUENCY_DIVIDE_64 = 0X03,
} my9291_cmd_frequency_t;

typedef enum my9291_cmd_scatter_t {
        MY9291_CMD_SCATTER_APDM = 0X00,
        MY9291_CMD_SCATTER_PWM = 0X01,
} my9291_cmd_scatter_t;

typedef struct {
        my9291_cmd_scatter_t scatter:1;
        my9291_cmd_frequency_t frequency:2;
        my9291_cmd_bit_width_t bit_width:2;
        my9291_cmd_reaction_t reaction:1;
        my9291_cmd_one_shot_t one_shot:1;
        unsigned char resv:1;
} __attribute__ ((aligned(1), packed)) my9291_cmd_t;

typedef struct {
    unsigned int red;
    unsigned int green;
    unsigned int blue;
    unsigned int white;
    unsigned int warm;
} my9291_color_t;

#define MY9291_COMMAND_DEFAULT { \
    .scatter = MY9291_CMD_SCATTER_APDM, \
    .frequency = MY9291_CMD_FREQUENCY_DIVIDE_1, \
    .bit_width = MY9291_CMD_BIT_WIDTH_8, \
    .reaction = MY9291_CMD_REACTION_FAST, \
    .one_shot = MY9291_CMD_ONE_SHOT_DISABLE, \
    .resv = 0 \
}

class my9291 {

    public:

        my9291(unsigned char di, unsigned char dcki, my9291_cmd_t command, unsigned char channels = 4);
        void setColor(my9291_color_t color);
        my9291_color_t getColor();
        void setState(bool state);
        bool getState();

    private:

        void _di_pulse(unsigned int times);
        void _dcki_pulse(unsigned int times);
        void _set_cmd(my9291_cmd_t command);
        void _send();
        void _write(unsigned int data, unsigned char bit_length);

        my9291_cmd_t _command;
        unsigned char _channels = 4;
        bool _state = false;
        my9291_color_t _color = {0, 0, 0, 0, 0};
        unsigned char _pin_di;
        unsigned char _pin_dcki;


};

#endif


src/my9291.cpp
/*

MY9291 LED Driver Arduino library 2.0.0
Based on the C driver by MaiKe Labs

Copyright (c) 2016 - 2026 MaiKe Labs
Copyright (c) 2017 - Xose Pérez (for the library)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include "my9291.h"

#if ARDUINO_ARCH_ESP8266

    extern "C" {
        void os_delay_us(unsigned int);
    }

#elif ARDUINO_ARCH_AVR

    #define os_delay_us delayMicroseconds

#endif

void my9291::_di_pulse(unsigned int times) {
	for (unsigned int i = 0; i < times; i++) {
		digitalWrite(_pin_di, HIGH);
		digitalWrite(_pin_di, LOW);
	}
}

void my9291::_dcki_pulse(unsigned int times) {
	for (unsigned int i = 0; i < times; i++) {
		digitalWrite(_pin_dcki, HIGH);
		digitalWrite(_pin_dcki, LOW);
	}
}

void my9291::_write(unsigned int data, unsigned char bit_length) {

    unsigned int mask = (0x01 << (bit_length - 1));

    for (unsigned int i = 0; i < bit_length / 2; i++) {
        digitalWrite(_pin_dcki, LOW);
        digitalWrite(_pin_di, (data & mask));
        digitalWrite(_pin_dcki, HIGH);
        data = data << 1;
        digitalWrite(_pin_di, (data & mask));
        digitalWrite(_pin_dcki, LOW);
        digitalWrite(_pin_di, LOW);
        data = data << 1;
    }

}

void my9291::_set_cmd(my9291_cmd_t command) {

	// ets_intr_lock();

    // TStop > 12us.
	os_delay_us(12);

    // Send 12 DI pulse, after 6 pulse's falling edge store duty data, and 12
	// pulse's rising edge convert to command mode.
	_di_pulse(12);

    // Delay >12us, begin send CMD data
	os_delay_us(12);

    // Send CMD data
    unsigned char command_data = *(unsigned char *) (&command);
    _write(command_data, 8);

	// TStart > 12us. Delay 12 us.
	os_delay_us(12);

    // Send 16 DI pulse，at 14 pulse's falling edge store CMD data, and
	// at 16 pulse's falling edge convert to duty mode.
	_di_pulse(16);

    // TStop > 12us.
	os_delay_us(12);

    // ets_intr_unlock();

}

void my9291::_send() {

	// Color to show
	unsigned int duty[6] = {0};
	if (_state) {
        if (_channels == 4) {
            duty[0] = _color.red;
            duty[1] = _color.green;
            duty[2] = _color.blue;
            duty[3] = _color.white;
            duty[4] = 0;
            duty[5] = 0;
        } else {
            duty[0] = _color.white;
            duty[1] = _color.warm;
            duty[2] = 0;
            duty[3] = _color.green;
            duty[4] = _color.red;
            duty[5] = _color.blue;
        }
    }

    unsigned char bit_length = 8;
	switch (_command.bit_width) {
		case MY9291_CMD_BIT_WIDTH_16:
			bit_length = 16;
			break;
		case MY9291_CMD_BIT_WIDTH_14:
			bit_length = 14;
			break;
		case MY9291_CMD_BIT_WIDTH_12:
			bit_length = 12;
			break;
		case MY9291_CMD_BIT_WIDTH_8:
			bit_length = 8;
			break;
		default:
			bit_length = 8;
			break;
	}

	// ets_intr_lock();

    // TStop > 12us.
	os_delay_us(12);

    // Send color data
    unsigned int length = (_channels == 4) ? 4 : 6;
    for (unsigned char channel = 0; channel < length; channel++) {
        _write(duty[channel], bit_length);
	}

	// TStart > 12us. Ready for send DI pulse.
	os_delay_us(12);

	// Send 8 DI pulse. After 8 pulse falling edge, store old data.
	_di_pulse(8);

	// TStop > 12us.
	os_delay_us(12);

	// ets_intr_unlock();

}

// -----------------------------------------------------------------------------

my9291_color_t my9291::getColor() {
    return _color;
}

void my9291::setColor(my9291_color_t color) {
    _color.red = color.red;
    _color.green = color.green;
    _color.blue = color.blue;
    _color.white = color.white;
    _color.warm = color.warm;
    DEBUG_MSG_MY9291("[MY9291] setColor: %d, %d, %d, %d, %d\n", _color.red, _color.green, _color.blue, _color.white, _color.warm);
    if (_state) _send();
}

bool my9291::getState() {
    return _state;
}

void my9291::setState(bool state) {
    _state = state;
    DEBUG_MSG_MY9291("[MY9291] setState: %s\n", _state ? "true" : "false");
    _send();
}

// -----------------------------------------------------------------------------

my9291::my9291(unsigned char di, unsigned char dcki, my9291_cmd_t command, unsigned char channels) {

	_pin_di = di;
	_pin_dcki = dcki;
    _command = command;
    if (channels == 4 || channels == 5) _channels = channels;

	pinMode(_pin_di, OUTPUT);
	pinMode(_pin_dcki, OUTPUT);

	digitalWrite(_pin_di, LOW);
	digitalWrite(_pin_dcki, LOW);

	// Clear all duty register
    if (_channels == 4) {
       	_dcki_pulse(32);
    } else {
       	_dcki_pulse(64);
    }

    // Send init command
	_set_cmd(command);

    DEBUG_MSG_MY9291("[MY9291] Initialized\n");

}


examples/uno/uno_rainbow.cpp
#include "my9291.h"

#define MY9291_DI_PIN       13
#define MY9291_DCKI_PIN     15
#define RAINBOW_DELAY       10

// MY9291 with 4 channels (like the AiThinker Ai-Light)
my9291 _my9291 = my9291(MY9291_DI_PIN, MY9291_DCKI_PIN, MY9291_COMMAND_DEFAULT, 4);

void rainbow(unsigned char index) {

    if (index < 85) {
        _my9291.setColor((my9291_color_t) { (unsigned int) index * 3, (unsigned int) 255 - index * 3, 0, 0, 0 });
    } else if (index < 170) {
        index -= 85;
        _my9291.setColor((my9291_color_t) { (unsigned int) 255 - index * 3, 0, (unsigned int) index * 3, 0, 0 });
    } else {
        index -= 170;
        _my9291.setColor((my9291_color_t) { 0, (unsigned int) index * 3, (unsigned int) 255 - index * 3, 0, 0 });
    }

}

void setup() {

    // Init serial port and clean garbage
    Serial.begin(115200);
    Serial.println();
    Serial.println();

    // MY9291
    _my9291.setState(true);

}

void loop() {

    static unsigned char count = 0;
    static unsigned long last = millis();
    if (millis() - last > RAINBOW_DELAY) {
        last = millis();
        rainbow(count++);
    }

}


examples/esp8285/esp8285_rainbow.cpp
#include <ESP8266WiFi.h>
#include <ESP8266mDNS.h>
#include <WiFiUdp.h>
#include <ArduinoOTA.h>
#include "credentials.h"
#include "my9291.h"

#define MY9291_DI_PIN       13
#define MY9291_DCKI_PIN     15
#define RAINBOW_DELAY       10

my9291 * _my9291;

void wifiSetup() {

    // Set WIFI module to STA mode
    WiFi.mode(WIFI_STA);

    // Connect
    Serial.printf("[WIFI] Connecting to %s\n", WIFI_SSID);
    WiFi.begin(WIFI_SSID, WIFI_PASS);

    // Wait
    while (WiFi.waitForConnectResult() != WL_CONNECTED) {
        Serial.println(" [WIFI] Connection Failed! Rebooting...");
        delay(5000);
        ESP.restart();
    }

    // Connected!
    Serial.printf("[WIFI] STATION Mode, SSID: %s, IP address: %s\n", WiFi.SSID().c_str(), WiFi.localIP().toString().c_str());

}

void rainbow(unsigned char index) {

    if (index < 85) {
        _my9291->setColor((my9291_color_t) { (unsigned int) index * 3, (unsigned int) 255 - index * 3, 0, 0, 0 });
    } else if (index < 170) {
        index -= 85;
        _my9291->setColor((my9291_color_t) { (unsigned int) 255 - index * 3, 0, (unsigned int) index * 3, 0, 0 });
    } else {
        index -= 170;
        _my9291->setColor((my9291_color_t) { 0, (unsigned int) index * 3, (unsigned int) 255 - index * 3, 0, 0 });
    }

}

void setup() {

    // Init serial port and clean garbage
    Serial.begin(115200);
    Serial.println();
    Serial.println();

    // WiFi
    wifiSetup();

    // OTA
    ArduinoOTA.onStart([]() {
        Serial.printf("[OTA] Start\n");
    });

    ArduinoOTA.onEnd([]() {
        Serial.printf("\n[OTA] End\n");
    });

    ArduinoOTA.onProgress([](unsigned int progress, unsigned int total) {
        Serial.printf("[OTA] Progress: %u%%\r", (progress / (total / 100)));
    });
    ArduinoOTA.onError([](ota_error_t error) {
        Serial.printf("\n[OTA] Error[%u]: ", error);
        if (error == OTA_AUTH_ERROR) Serial.printf("Auth Failed\n");
        else if (error == OTA_BEGIN_ERROR) Serial.printf("Begin Failed\n");
        else if (error == OTA_CONNECT_ERROR) Serial.printf("Connect Failed\n");
        else if (error == OTA_RECEIVE_ERROR) Serial.printf("Receive Failed\n");
        else if (error == OTA_END_ERROR) Serial.printf("End Failed\n");
    });
    ArduinoOTA.begin();

    // MY9291 with 4 channels (like the AiThinker Ai-Light)
    _my9291 = new my9291(MY9291_DI_PIN, MY9291_DCKI_PIN, MY9291_COMMAND_DEFAULT, 4);
    _my9291->setState(true);

}

void loop() {

    ArduinoOTA.handle();

    static unsigned char count = 0;
    static unsigned long last = millis();
    if (millis() - last > RAINBOW_DELAY) {
        last = millis();
        rainbow(count++);
    }

}
